### PR TITLE
CORS-3251: Create capz machine manifests

### PR DESCRIPTION
**Pull request by rna-afk**
Creating machine manifests for azure CAPI implementation.

### CODE REVIEW

File: pkg/asset/machines/azure/azuremachines.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	}
+
+	providerID := fmt.Sprintf("%s-identity", clusterID)
+	var result []*asset.RuntimeFile
+	for idx := int64(0); idx < total; idx++ {
+		zone := mpool.Zones[int(idx)%len(mpool.Zones)]
+
+		networkInterfaces := []capz.NetworkInterface{{
+			SubnetName:            subnet,
+			AcceleratedNetworking: additionalCapabilities.UltraSSDEnabled,
```
Comment (vincepri) [inline pkg/asset/machines/azure/azuremachines.go:114]: Is `UltraSSDEnabled` in the correct place here? Seems that one is related to disks, rather than network interfaces?

(not super familiar with Azure specifically, was going by https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.management.compute.models.additionalcapabilities.ultrassdenabled?view=azure-dotnet-legacy)

**Comment by patrickdillon** [comment]
oh cool, with the provider in place we might be able to test this

/test e2e-azure-capi-ovn

File: pkg/asset/machines/azure/azuremachines.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		OSType:     "Linux",
+		DiskSizeGB: &mpool.DiskSizeGB,
+		ManagedDisk: &capz.ManagedDiskParameters{
+			StorageAccountType: mpool.DiskType,
+		},
+	}
+	ultrassd := mpool.UltraSSDCapability == "Enabled"
+	additionalCapabilities := capz.AdditionalCapabilities{
+		UltraSSDEnabled: &ultrassd,
+	}
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:84]: Generally the API is expecting these to be pointers, which you are solving by using `DeepCopy`. Instead, create the variables as pointers and drop DeepCopy:
```suggestion
	additionalCapabilities := &capz.AdditionalCapabilities{
		UltraSSDEnabled: &ultrassd,
	}
```

File: pkg/asset/machines/azure/azuremachines.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+	diagnostics := capz.Diagnostics{
+		Boot: nil,
+	}
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:88]: You could just drop this entirely since it's empty.

File: pkg/asset/machines/azure/azuremachines.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+	providerID := fmt.Sprintf("%s-identity", clusterID)
+	var result []*asset.RuntimeFile
+	for idx := int64(0); idx < total; idx++ {
+		zone := mpool.Zones[int(idx)%len(mpool.Zones)]
+
+		networkInterfaces := []capz.NetworkInterface{{
+			SubnetName:            subnet,
+			AcceleratedNetworking: &isAcceleratedNetworkEnabled,
+		}}
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:117]: in testing, these network interfaces were problematic. They are optional, so let's just remove this from the PR, and we could add them in as we iterate.

File: pkg/asset/machines/azure/azuremachines.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+				SSHPublicKey:           "",
+				AdditionalTags:         tags,
+				AdditionalCapabilities: additionalCapabilities,
+				AllocatePublicIP:       false,
+				EnableIPForwarding:     false,
+			},
+		}
+		*azureMachine.Spec.Image = *image
+		*azureMachine.Spec.AdditionalCapabilities = *additionalCapabilities
+		*azureMachine.Spec.SecurityProfile = *securityProfile
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:133]: Let's move these into the literal, so they are set like the other fields immediately above this

File: pkg/asset/machines/azure/azuremachines.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+			Labels: map[string]string{
+				"cluster.x-k8s.io/control-plane": "",
+				"install.openshift.io/bootstrap": "",
+				"cluster.x-k8s.io/cluster-name":  clusterID,
+			},
+		},
+		Spec: capz.AzureMachineSpec{
+			ProviderID:         ptr.To(providerID),
+			VMSize:             mpool.InstanceType, // required
+			Image:              image.DeepCopy(),
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:182]:
```suggestion
			Image:              image,
```

File: pkg/asset/machines/azure/azuremachines.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+			OSDisk:             osDisk, // required
+			SSHPublicKey:       "",
+			AdditionalTags:     tags,
+			AllocatePublicIP:   true,
+			EnableIPForwarding: false,
+		},
+	}
+	*bootstrapAzureMachine.Spec.Image = *image
+	*bootstrapAzureMachine.Spec.AdditionalCapabilities = *additionalCapabilities
+	*bootstrapAzureMachine.Spec.SecurityProfile = *securityProfile
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:193]: ditto. Set these in the literal

File: pkg/asset/machines/azure/azuremachines.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+		Spec: capz.AzureMachineSpec{
+			ProviderID:         ptr.To(providerID),
+			VMSize:             mpool.InstanceType, // required
+			Image:              image.DeepCopy(),
+			FailureDomain:      ptr.To(mpool.Zones[0]),
+			OSDisk:             osDisk, // required
+			SSHPublicKey:       "",
+			AdditionalTags:     tags,
+			AllocatePublicIP:   true,
+			EnableIPForwarding: false,
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:188]: nit: let's remove the fields that are using the default values, so drop `EnableIPForwarding` and `SSHPublicKey`

File: pkg/asset/machines/azure/azuremachines.go (hunk 5 above)
Comment (rna-afk) [reply pkg/asset/machines/azure/azuremachines.go:133]: Wouldn't the mutable nature cause problems? I had some problems back in the CPMS manifests creation, hence the DeepCopy before although I don't see them changing. Or are you suggesting something else?

Comment (patrickdillon) [inline pkg/asset/machines/azure/azuremachines.go:133]:
```suggestion
			Spec: capz.AzureMachineSpec{
				ProviderID:             ptr.To(providerID),
				VMSize:                 mpool.InstanceType, // required
				FailureDomain:          ptr.To(zone),
				Image:                  image
				OSDisk:                 osDisk, // required
				SSHPublicKey:           "",
				AdditionalTags:         tags,
				AdditionalCapabilities: additionalCapabilities,
				AllocatePublicIP:       false,
				EnableIPForwarding:     false,
				SecurityProfile:        securityProfile,
			},
		}
```

Comment (patrickdillon) [reply pkg/asset/machines/azure/azuremachines.go:133]: Shouldn't be a problem. My suggestion should be the equivalent of what you're doing now, except instead of setting the values in two different ways, all of the values are being set in the [struct literal](https://go.dev/tour/moretypes/5). I tried to illustrate the idea here: https://github.com/openshift/installer/pull/7969/files#r1488727469

Comment (rna-afk) [reply pkg/asset/machines/azure/azuremachines.go:133]: Ok it's the same thing. We had a problem in CPMS where since all these variables have the same address, change in one of these resulted in all of them changing so we had to use DeepCopy/*var=*var. The change in value happened in the CPMS operator and if the CAPI operator changed this, it would be problematic.

File: pkg/infrastructure/platform/platform.go [hunk 9]
```diff
@@ -53,6 +54,9 @@ func ProviderForPlatform(platform string, fg featuregates.FeatureGate) (infrastr
 		}
 		return terraform.InitializeProvider(aws.PlatformStages), nil
 	case azuretypes.Name:
+		if fg.Enabled(configv1.FeatureGateClusterAPIInstall) {
+			return clusterapi.InitializeProvider(azureinfra.InitializeProvider()), nil
+		}
```
Comment (patrickdillon) [inline pkg/infrastructure/platform/platform.go:59]: Add this same code to the `platform_altinfra.go` file

That altinfra file is selected using build tags, and we use it to produce a terraform-free installer binary (in a separate image).

File: pkg/infrastructure/azure/azure.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+// const (
+// 	tfVarsFileName         = "terraform.tfvars.json"
+// 	tfPlatformVarsFileName = "terraform.platform.auto.tfvars.json"
+// 	clusterOutputFileName  = "cluster.azuresdk.vars.json"
+
+// 	defaultDescription = "Created by Openshift Installer"
+// 	ownedTagKey        = "kubernetes.io/cluster/%s"
+// 	ownedTagValue      = "owned"
+// )
```
Comment (patrickdillon) [inline pkg/infrastructure/azure/azure.go:15]: delete this

File: pkg/infrastructure/azure/azure.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+// InitializeProvider initializes the Azure SDK provider.
+func InitializeProvider() clusterapi.Provider {
+	return &InfraProvider{}
+}
+
+// Name returns the name of the provider.
+func (p *InfraProvider) Name() string {
+	return "azure"
+}
```
Comment (patrickdillon) [inline pkg/infrastructure/azure/azure.go:28]:
```suggestion
var _ clusterapi.Provider = (*Provider)(nil)

// Provider implements Azure CAPI installation.
type Provider struct{}

// Name gives the name of the provider, Azure.
func (*Provider) Name() string { return azuretypes.Name }
```

Comment (patrickdillon) [reply pkg/infrastructure/azure/azure.go:28]: introduction of the azure capi provider should really be in its own commit

File: pkg/asset/manifests/azure/cluster.go [hunk 12]
```diff
@@ -40,7 +42,7 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
 			Namespace: capiutils.Namespace,
 		},
 		Spec: capz.AzureClusterSpec{
-			ResourceGroup: clusterID.InfraID,
+			ResourceGroup: fmt.Sprintf("%s-rg", clusterID.InfraID),
```
Comment (patrickdillon) [inline pkg/asset/manifests/azure/cluster.go:45]: this lgtm, but should be in its own commit (it is a standalone fix unrelated to the capz machine changes)

File: go.mod [hunk 13]
```diff
@@ -109,7 +109,7 @@ require (
 	k8s.io/utils v0.0.0-20240102154912-e7106e64919e
 	sigs.k8s.io/cluster-api v1.5.3
 	sigs.k8s.io/cluster-api-provider-aws/v2 v2.0.0-00010101000000-000000000000
-	sigs.k8s.io/cluster-api-provider-azure v0.0.0-00010101000000-000000000000
+	sigs.k8s.io/cluster-api-provider-azure v1.13.0
```
Comment (patrickdillon) [inline go.mod:112]: the vendor changes should be in their own commit

**Review by patrickdillon — COMMENTED** [review]
This looks like it's in good shape but just needs several house keeping items:

- [ ] remove wip from title (replace with jira) 
- [ ] would be ideal to restructure the commits
- [ ] linter wants you to use a switch 
- [ ] empty provider needs to be refactored a bit

note that the linter wants

**Comment by patrickdillon** [comment]
/approve

/test altinfra-e2e-azure-capi-ovn

This is looking really good.

File: pkg/asset/machines/azure/azuremachines.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+		userTagKeys = append(userTagKeys, key)
+	}
+	sort.Strings(userTagKeys)
+
+	for _, k := range userTagKeys {
+		if forbiddenTags.Has(k) {
+			return nil, fmt.Errorf("user tags may not clobber %s", k)
+		}
+		tags[k] = usertags[k]
+	}
```
Comment (r4f4) [inline pkg/asset/machines/azure/azuremachines.go:245]: Why not use a set here and then report all the conflicting tags at once? If a user has multiple conflicting tags, they'd end up running the installer multiple times before they get rid of all conflicts
```suggestion
	userTagKeys := sets.New[string]()
	for key := range usertags {
		userTagKeys.Insert(key)
	}

    if clobberedTags := userTagKeys.Intersection(forbiddenTags); clobberedTags.Len() > 0 {
		return nil, fmt.Errorf("user tag keys %v are not allowed", sets.List(clobberedTags)) // maybe not needed to sort the keys here; it's only useful for comparison in unit tests
	}
	for _, k := range sets.List(userTagKeys) {
		tags[k] = usertags[k]
	}
```

Comment (rna-afk) [reply pkg/asset/machines/azure/azuremachines.go:245]: I just picked the code from the installer for this one.
https://github.com/openshift/installer/blob/b8d567115ff6a93ad048808fa4c8a52eb78fe954/pkg/asset/machines/aws/awsmachines.go#L129

Comment (r4f4) [reply pkg/asset/machines/azure/azuremachines.go]: Aha, I'll send a PR.

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by patrickdillon** [comment]
These tests are permafailing. Installs look fine
